src/server/conn.rs

```diff
@@ -114,45 +141,160 @@ impl Conn {
               S: StoreAddrResolver
     {
-        let mut bufs = vec![];
+        match self.conn_type {
+            ConnType::HandShake => self.handshake(event_loop),
+            ConnType::Rpc => self.read_rpc(event_loop),
+            ConnType::Snapshot => self.read_snapshot(event_loop),
+        }
+    }
 
-        loop {
-            // Because we use the edge trigger, so here we must read whole data.
-            if self.payload.is_none() {
-                try!(try_read_data(&mut self.sock, &mut self.header));
-                if self.header.remaining() > 0 {
-                    // we need to read more data for header
-                    break;
+    fn handshake<T, S>(&mut self, event_loop: &mut EventLoop<Server<T, S>>) -> Result<Vec<ConnData>>
+        where T: RaftStoreRouter,
+              S: StoreAddrResolver
+    {
+        let msg_or_none = try!(self.read_one_message());
+        if msg_or_none.is_none() {
+            return Ok(vec![]);
+        }
+
+        let mut msg = msg_or_none.unwrap();
+        match msg.get_msg_type() {
+            MessageType::Snapshot => {
+                let mut worker = Worker::new("snapshot receiver".to_owned());
+                // TODO we need store id here!!
+                let runner = Runner::new(&self.snap_path,
+                                         msg.take_snapshot_file(),
+                                         msg.take_raft(),
+                                         self.last_msg_id,
+                                         self.token,
+                                         self.ch.clone());
+                try!(worker.start(runner));
+                debug!("receive a snapshot connection\n");
+                self.snapshot_receiver = Some(SnapshotReceiver {
+                    buf: vec![0; 10 * (1 << 20)],
+                    pos: 0,
+                    worker: worker,
+                    more: false,
+                    file_size: 0,
+                    read_size: 0,
+                });
+                self.conn_type = ConnType::Snapshot;
+                self.read_snapshot(event_loop)
+            }
+            _ => {
+                self.conn_type = ConnType::Rpc;
+                let mut first = vec![ConnData {
+                                         msg_id: self.last_msg_id,
+                                         msg: msg,
+                                     }];
+                let mut rem = try!(self.read_rpc(event_loop));
+                first.append(&mut rem);
+                Ok(first)
+            }
+        }
+    }
+
+    fn read_snapshot<T, S>(&mut self, _: &mut EventLoop<Server<T, S>>) -> Result<Vec<ConnData>>
+        where T: RaftStoreRouter,
+              S: StoreAddrResolver
+    {
+        if let Some(ref mut receiver) = self.snapshot_receiver {
+            let (mut pos, len) = (receiver.pos, receiver.buf.len());
+            loop {
+                match try!(self.sock.try_read(&mut receiver.buf[pos..len])) {
```

Comment:
I am very confused that why using payload directly, the logic is same, so I don't know why it doesn't work.